### PR TITLE
Add 2 new tests to test/preparation/test_protonate.py

### PR DESCRIPTION
**Pull request by SwordJack**
Dear colleagues,

Concerning the issue#156, I added 2 new unit tests in `test/preparation/test_protonate.py` on `_fix_pybel_output` and `protonate_ligand_with_pybel` in `preparation.protonate`.

By the way, an update to the `environment.yml` to improve the convenience of environment configuration is also committed.

Best,
Yinjie.

### CODE REVIEW

File: test/preparation/test_protonate.py [hunk 1]
```diff
@@ -277,3 +321,28 @@ def test_protonate_stru_imputed():
     prot.protonate_stru(stru)
 
     assert stru
+
+def test_pdb2pqr_pybel_protonate_pdb_7SI9():
```
Comment (shaoqx) [inline test/preparation/test_protonate.py:325]: remove this

File: test/preparation/test_protonate.py [hunk 2]
```diff
@@ -277,3 +321,28 @@ def test_protonate_stru_imputed():
...
+
+    prot.protonate_stru(stru=stru)
+
+    # prot.protonate_peptide_with_pdb2pqr(stru=stru, int_pqr_path=int_pqr_path, int_pdb_path=int_pdb_path)
+    # prot.protonate_ligand_with_pybel(stru=stru, int_ligand_file_dir=int_ligand_file_dir)
+
+    sp.save_structure(outfile=out_pdb_file, stru=stru)
+
+
+if __name__ == '__main__':
```
Comment (shaoqx) [inline test/preparation/test_protonate.py:346]: remove this

**Review by shaoqx — CHANGES_REQUESTED** [review]
See comments

File: test/preparation/data/7SI9_rm_water_disconnected_rmH_aH_EA166Q_ff_min_rmW_rmH.pdb
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/data/7SI9_rm_water_disconnected_rmH_aH_EA166Q_ff_min_rmW_rmH.pdb:1]: Done.

File: test/preparation/data/ligand_test_4WI_pybel_badname.pdb
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/data/ligand_test_4WI_pybel_badname.pdb:1]: Done.

File: test/preparation/test_protonate.py [hunk 3]
```diff
@@ -197,6 +198,23 @@ def test_protonate_ligand_with_pybel():
...
+    pdb_file = f"{DATA_DIR}/{file_prefix}.pdb"
+    int_ligand_file_dir=f"{WORK_DIR}"
+    out_pdb_file = f"{WORK_DIR}/{file_prefix}_stru_output.pdb"
+    
+    stru = sp.get_structure(pdb_file)
+    prot.protonate_ligand_with_pybel(stru=stru, int_ligand_file_dir=int_ligand_file_dir)
+    
+    for atom in stru.ligands[0].atoms:
+        element_from_name = atom.name[:len(atom.element)]   # The name of the atom should start with the element symbol.
+        # print(f'Name: {element_from_name}; Element: {atom.element}')
```
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/data/ligand_test_4WI.pdb
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/data/ligand_test_4WI.pdb:1]: Done.

File: test/preparation/test_protonate.py [hunk 4]
```diff
@@ -197,6 +198,23 @@ def test_protonate_ligand_with_pybel():
 
     assert not os.path.isdir(int_lig_dir)
 
+def test_protonate_ligand_with_pybel_7si9():
+    """Test the protonation to ligands in a structure instance"""
+    # _LOGGER.setLevel(0)
```
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/test_protonate.py [hunk 5]
```diff
@@ -247,6 +265,17 @@ def test_pybel_protonate_pdb_ligand_4CO():
     # fs.safe_rm(out_ligand_path)
     # assert not os.path.isdir(out_ligand_path)
 
+def test_pybel_protonate_pdb_ligand_4WI():
```
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/test_protonate.py (hunk 1 above)
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/test_protonate.py (hunk 2 above)
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/test_protonate.py [hunk 6]
```diff
@@ -247,6 +265,17 @@ def test_pybel_protonate_pdb_ligand_4CO():
...
+    """Test ligand protonation (pybel)."""
+
+    _LOGGER.setLevel(0)
```
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

File: test/preparation/test_protonate.py [hunk 7]
```diff
@@ -262,6 +291,21 @@ def test_fix_pybel_output():
     fs.safe_rm(out_ligand_path)
     assert not os.path.isdir(out_ligand_path)
 
+def test_fix_pybel_output_4WI(caplog):
+    """Test protonated ligand (pybel) name fixing."""
+    # _LOGGER.setLevel(0)
```
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/test_protonate.py]: Done.

**Comment by SwordJack** [comment]
@shaoqx  All the conversations are resolved.

File: enzy_htp/structure/structure_operation/general.py [hunk 8]
```diff
@@ -49,6 +49,9 @@ def remove_non_peptide(stru: Structure) -> Structure:
         ch.delete_from_parent()
     return stru
 
+def remove_hydrogens(ligand: Ligand, polypeptide_only: bool) -> None:
```
Comment (shaoqx) [inline enzy_htp/structure/structure_operation/general.py:52]: Please make it general to Residues and dont put these fix in the unit test PR

File: enzy_htp/structure/residue.py [hunk 9]
```diff
@@ -301,6 +301,10 @@ def is_connected(self) -> bool:
     def has_init_charge(self) -> bool:
         """check if self has charge"""
         return math.prod([atom.has_init_charge() for atom in self.atoms])  
+    
+    def has_hydrogen(self) -> bool:
```
Comment (shaoqx) [inline enzy_htp/structure/residue.py:305]: We try to make each PR closely bind to its theme. So please have these fix in a different PR

**Review by shaoqx — CHANGES_REQUESTED** [review]
Please change according to the comments

File: enzy_htp/structure/structure_operation/general.py (hunk 8 above)
Comment (SwordJack) [reply enzy_htp/structure/structure_operation/general.py]: The modifications to files in `enzyhtp` directory are now unstaged.

File: enzy_htp/structure/residue.py (hunk 9 above)
Comment (SwordJack) [reply enzy_htp/structure/residue.py]: The modifications to files in `enzyhtp` directory are now unstaged.

File: test/preparation/data/7SI9_E166Q_protonated.pdb
[the remark this answers is not in the document]
Comment (SwordJack) [reply test/preparation/data/7SI9_E166Q_protonated.pdb:1]: OK.

**Review by shaoqx — APPROVED** [review]
Good job making the unit tests!

**Pull request merged**